### ISSUE TITLE
Parallel deploys not queueing consistently

### ISSUE THREAD

**Issue by ryanblock** [arc-repos/arc-workflows#145]
If too many `UpdateFunctionCode` operations are performed at once, Lambda will error and rate limit you.

From #110, we created `PARALLEL_DEPLOYS_PER_SECOND`, which can be set by users. The default value is `8`.

However, the following behavior can be observed seen during deploys of large quantities of Lambdas:
- `deploy` correctly [chunks up the set of functions](https://github.com/arc-repos/arc-workflows/blob/master/src/deploy/_chunk.js) to be deployed
- Somewhere something goes wrong, and `deploy` attempt dozens of `UpdateFunctionCode` operations per second, irrespective of the `PARALLEL_DEPLOYS_PER_SECOND` value

Example `UpdateFunctionCode` operations I performed on a > 60 Lambda project today:

`PARALLEL_DEPLOYS_PER_SECOND` = 5
- `2018-09-27T21:15:14`: 5 `UpdateFunctionCode` calls during this second
- `2018-09-27T21:15:15`: 24 `UpdateFunctionCode` calls during this second
- `2018-09-27T21:15:16`: 20 `UpdateFunctionCode` calls during this second
- `2018-09-27T21:15:17`: 7 `UpdateFunctionCode` calls during this second
- `2018-09-27T21:15:18`: 4 `UpdateFunctionCode` calls during this second
- `2018-09-27T21:15:19`: 4 `UpdateFunctionCode` calls during this second

`PARALLEL_DEPLOYS_PER_SECOND` = 10
- `2018-09-27T21:12:01`: 27 `UpdateFunctionCode` calls during this second
- `2018-09-27T21:12:02`: 23 `UpdateFunctionCode` calls during this second
- `2018-09-27T21:12:03`: 12 `UpdateFunctionCode` calls during this second
- [errors]

### Repro steps
Because the part of the deploy operation that will result in AWS errors is `UpdateFunctionCode`, I measured this by logging a timestamp within the `_upload` function within https://github.com/arc-repos/arc-workflows/blob/master/src/deploy/lambda/05-upload-zip.js

After that, run a full deploy of a large, many-Lambda project.

**Comment by ryanblock** [comment]
Finding: parallel deploys weren't exactly parallel due. This led to bunching up of certain operations, resulting in erratic flooding of Lambda `UpdateFunctionCode` requests.

More specifically:
There are 7 operations to deploying a Lambda:
1. verify lambda - completes nearly instantly
2. pre-deploy plugins - not in use by us rn, so it's skipped
3. install modules - completes wildly inconsistently for a variety of reasons:
  - some modules may be locally cached, while others are not
  - some functions may have many more modules, or modules with many more dependencies than others
  - npm stuff
4. copy shared - completes fairly quickly, but more importantly is consistent across all Lambdas
5. upload zip - completes fairly inconsistently, as it's actually two operations: compressing the function (which have large variation in number of modules files, etc.), and uploading (which will have variation in size to be uploaded)
6. post-deploy plugins - not in use by us rn, so it's skipped
7. done - progress, error reporting, etc.

So basically although we're starting these operations in chunks of _n_ once per second, we're still creating a race where many modules from many starting points wind up running `UpdateFunctionCode` at the same time (hence the wacky distribution above).

My solution for this was to break every deploy into two macro steps:
1. steps 1-4, fully parallelized, no queueing
2. steps 5-7, parallelized across queueing chunks

Tested this a bunch last night and I am no longer seeing `UpdateFunctionCode` rate limiting errors.

**CAVEAT**
This does not create a perfectly shaped stream of deployment transactions per second, because there's still a race present. Can you find it?

Ok, I'll share: we are queueing deploy ops start each second, and not waiting for the last round of ops to finish before starting the next batch (which is probably the right approach!).

However, in operation 5, the moment we will now begin queuing, function compression starts the race again. Granted, this time it's much more isolated, because all the really gnarly filesystem operations are already done by this moment. But because not all Lambdas are the same size, compressing each function takes a variable amount of time, and that is made a bit worse by the variation in upload time, too.

Therefore, we'll still see some `UpdateFunctionCode` operation bunching with an imperfect number of transactions per second. But the good news is they appear to be much more consistent now, with significantly less variation per second. While I am still seeing `UpdateFunctionCode` operations bunch up a bit, now it's far enough below the Lambda control backplane rate limiting threshold that I've yet to see a single error.

---

**Room for improvement?**

I considered alternatives to help us to minimize or eliminate the compression race as well, but they're all crappy and inefficient by my view.

The best example I can see would be:
- Before the queue processing begins, write all functions' zips as a file to disk
- Then, as queue processing is about to begin, read all those zip files back into memory
- Then begin draining the queue, which would ensure perfect target `UpdateFunctionCode` tps

However, that means slowing everything else down significantly by reading all function contents, writing them to disk, then reading them again (i.e. adding a full write to disk and read from disk from the entire codebase).

This could be sidestepped by reading all the zips directly into memory and only starting the queue processing after that is complete, but now we're loading the entire codebase + modules into memory, which is almost certain to cause problems for larger projects (as opposed to what we do now, which is load just a few functions into memory at a time, since zipping is part of what's queued).

Open to ideas though!
